File: Array/704.c
#include<stdio.h>
#include<string.h>

//704.Binary search

int search(int* nums, int numsSize, int target) {
    int left = 0;
    int right = numsSize-1;
    int middle = 0;

    while(left <= right){
        middle = (left+right)/2;
        if(nums[middle] < target){
            left = middle+1;
        }else if(nums[middle] > target){
            right = middle -1;
        }else{
            return middle;
        }
    }
    //沒找到
    return -1;
}


int main(){
    int nums[] = {-1,0,3,5,9,12};
    int numsSize = sizeof(nums)/sizeof(nums[0]);
    int target = 12;
    printf("%d", search(nums, numsSize, target));
    return 0;
}

File: Array/704.py
def search(self, nums, target):

    left = 0
    right = len(nums)-1
    
    while left <= right:
        mid = (left + right)//2
        if nums[mid] == target:
            return mid
        elif nums[mid] > target:
            right = mid-1
        else:
            left = mid +1
    return -1


def run_test(nums, target, expected):
    ans = search(nums, target)
    print(f"nums={nums}, target={target}")
    print(f"  ans={ans}, expected={expected}  --> {'PASS' if ans == expected else 'FAIL'}")
    if ans != expected:
        raise AssertionError("Test failed")


if __name__ == "__main__":
    run_test([-1,0,3,5,9,12], 9, 4)
    run_test([-1,0,3,5,9,12], 2, -1)
    run_test([1], 1, 0)
    run_test([1], 0, -1)
    run_test([1,2,3,4,5], 1, 0)
    run_test([1,2,3,4,5], 5, 4)
    run_test([1,2,3,4,5], 3, 2)
    print("All tests passed!")
